File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/weapon/ViperManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.weapon;

import net.igneo.icv.ICV;
import net.igneo.icv.enchantment.EnchantType;
import net.igneo.icv.enchantmentActions.EntityTracker;
import net.igneo.icv.entity.ICVEntity;
import net.igneo.icv.entity.ModEntities;
import net.igneo.icv.init.ICVUtils;
import net.igneo.icv.networking.ModMessages;
import net.igneo.icv.networking.packet.GustS2CPacket;
import net.igneo.icv.sound.ModSounds;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec3;

import java.util.List;


public class ViperManager extends WeaponEnchantManager implements EntityTracker {

  ICVEntity child;
  boolean nullCheck = false;

  public ViperManager(Player player) {
    super(EnchantType.WEAPON, player, new ResourceLocation(ICV.MOD_ID, "dual_handed_slash_cross"));
  }
  
  @Override
  public void activate() {
    super.activate();
    if (player.level() instanceof ServerLevel level) {
      child = ModEntities.SNAKE_BITE.get().create(player.level());
      child.setOwner(player);
      child.setPos(player.getEyePosition().subtract(0,1,0));
      child.setDeltaMovement(ICVUtils.getFlatDirection(player.getYRot(),1,0));
      level.addFreshEntity(child);
      syncClientChild((ServerPlayer) player,child,this);
      nullCheck = true;
    }
  }

  @Override
  public float getDamageBonus() {
    if (target == null) {
      return 0;
    } else if (target instanceof LivingEntity entity) {
      if (!entity.getActiveEffects().isEmpty()) {
        for (MobEffectInstance instance : entity.getActiveEffects()) {
          MobEffect effect = instance.getEffect();
          if (!effect.isBeneficial()) {
            int duration = instance.getDuration() - 20;
            entity.removeEffect(effect);
            if (duration > 0) {
              entity.addEffect(new MobEffectInstance(effect,duration,instance.getAmplifier(),instance.isAmbient(),instance.isVisible()));
            }
          }
        }
        return 4;
      }
    }
    return 0;
  }

  @Override
  public boolean canUse() {
    return super.canUse() && child == null;
  }

  @Override
  public ICVEntity getChild() {
    return child;
  }

  @Override
  public void setChild(ICVEntity entity) {
    child = entity;
  }

  @Override
  public void tick() {
    super.tick();
    if (nullCheck && player instanceof ServerPlayer serverPlayer && child == null) {
      syncClientChild(serverPlayer,null,this);
      nullCheck = false;
    }
  }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/weapon/ViperManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.weapon

import net.igneo.icv.ICV
import net.igneo.icv.enchantment.EnchantType
import net.igneo.icv.enchantmentActions.EntityTracker
import net.igneo.icv.entity.ICVEntity
import net.igneo.icv.entity.ModEntities
import net.igneo.icv.init.ICVUtils
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player

class ViperManager(player: Player?) :
    WeaponEnchantManager(EnchantType.WEAPON, player, ResourceLocation(ICV.MOD_ID, "dual_handed_slash_cross")),
    EntityTracker {
    override var child: ICVEntity? = null
    private var nullCheck: Boolean = false

    override fun activate() {
        super.activate()
        val level = player!!.level()
        if (player!!.level() is ServerLevel) {
            child = ModEntities.SNAKE_BITE.get().create(player!!.level())
            child!!.owner = player
            child!!.setPos(player!!.eyePosition.subtract(0.0, 1.0, 0.0))
            child!!.deltaMovement = ICVUtils.getFlatDirection(player!!.yRot, 1f, 0.0)
            level.addFreshEntity(child)
            syncClientChild(player as ServerPlayer, child, this)
            nullCheck = true
        }
    }

    override val damageBonus: Float
        get() {
            val living = target as? LivingEntity ?: return 0f

            if (living.activeEffects.isNotEmpty()) {
                for (instance in living.activeEffects) {
                    val effect = instance.effect
                    if (!effect.isBeneficial) {
                        val duration = instance.duration - 20
                        living.removeEffect(effect)
                        if (duration > 0) {
                            living.addEffect(
                                MobEffectInstance(
                                    effect,
                                    duration,
                                    instance.amplifier,
                                    instance.isAmbient,
                                    instance.isVisible
                                )
                            )
                        }
                    }
                }
                return 4f
            }

            return 0f
        }


    override fun canUse(): Boolean {
        return super.canUse() && child == null
    }

    override fun tick() {
        super.tick()
        if (nullCheck && player is ServerPlayer && child == null) {
            syncClientChild(player as ServerPlayer, null, this)
            nullCheck = false
        }
    }
}